File: 20190519python/0430. Flatten a Multilevel Doubly Linked List.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""
class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        if not head:
            return head
        node=Node(0,None,None,None)
        node.next=head
        p=node
        list_stack=[]
        while list_stack or p.next:
            if not p.next:
                curList=list_stack.pop()
                p.next=curList
                curList.prev=p
            p=p.next
            while p.next or p.child:
                if p.child:
                    if p.next:
                        list_stack.append(p.next)
                    p.child.prev=p
                    p.next=p.child
                    p.child=None
                    p=p.next
                elif not p.child and p.next:
                    p=p.next
        return node.next

File: archives/C#/0430. Flatten a Multilevel Doubly Linked List.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;

    public Node(){}
    public Node(int _val,Node _prev,Node _next,Node _child) {
        val = _val;
        prev = _prev;
        next = _next;
        child = _child;
}
*/
public class Solution {
    public Node Flatten(Node head) {
        if(head==null){
            return head;
        }
        Stack<Node> rep=new Stack<Node>();
        rep.Push(head);
        
        Node node=new Node(0,null,null,null);
        Node p=node;
        while(rep.Count()!=0){
            Node tmp=rep.Pop();
            tmp.prev=p;
            p.next=tmp;
            p=p.next;
            
            while(p.next!=null || p.child!=null){
                if(p.child!=null){
                    if(p.next!=null){
                        rep.Push(p.next);
                    }
                    p.next=p.child;
                    p.child.prev=p;
                    p.child=null;
                }
                p=p.next;
            }            
        }
        node.next.prev=null;
        return node.next;
    }
}
